mypy/semanal.py

```diff
@@ -534,7 +541,7 @@ def visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
                 assert defn.impl is defn.items[-1]
                 defn.items = defn.items[:-1]
             elif not self.is_stub_file and not non_overload_indexes:
-                if not (self.is_class_scope() and self.type.is_protocol):
+                if not (self.type and not self.is_func_scope() and self.type.is_protocol):
```

Comment:
Yes, I also noticed this would be very useful to somehow "remember" the bindings. For simple cases I was thinking about something like inlining methods (very approximately):
```python
    @inline
    def meth(self) -> bool:
        return self.x is not None and self.y is not None
    if self.meth(): # use original expression returned by 'meth' here
        ...
```